improve scratch pad story

pfitzseb opened the pull request:
This improves the scratch pad/storage story a bit by
- adding a `@with` macro for interactive use,
- changing to a `Infiltrator.store` global to access the store (with proper autocompletion),
- and exposing `set_store(Main)` for the Exfiltrate behavior.

cc @antoine-levitt, @pdeffebach

antoine-levitt commented on README.md:
```diff
@@ -31,9 +31,23 @@ Assigns all local variables into the scratch pad.
-and reset it with `clear_scratch_pad`.
+Exfiltrating variables (with `@exfiltrate` or by assignment in a `@infiltrate` session) happens by
+assigning the variable to a global scratch pad (backed by a module). You can access this store with
+`Infiltrator.store`; any exfiltrated objects can be directly accessed, e.g. via `Infilatrator.store.mysymbol`.
```
myvariable would be clearer for a user

antoine-levitt commented on README.md:
```diff
@@ -31,9 +31,23 @@ Assigns all local variables into the scratch pad.
+assigning the variable to a global scratch pad (backed by a module). You can access this store with
+`Infiltrator.store`; any exfiltrated objects can be directly accessed, e.g. via `Infilatrator.store.mysymbol`.
+
+You can reset the module with `Infiltrator.clear_store` or assign a specific module with `Infiltrator.set_store`.
```
parenthesis on function call?

antoine-levitt commented on README.md:
```diff
@@ -31,9 +31,23 @@ Assigns all local variables into the scratch pad.
 Originally implemented and named by Antoine Levitt in [Exfiltrator.jl](https://github.com/antoine-levitt/Exfiltrator.jl).
 
 ## The scratch pad
```
Remove all ref to scratch in favor of store?

antoine-levitt commented on README.md:
```diff
@@ -31,9 +31,23 @@ Assigns all local variables into the scratch pad.
 Originally implemented and named by Antoine Levitt in [Exfiltrator.jl](https://github.com/antoine-levitt/Exfiltrator.jl).
```
I appreciate the credit (especially for the name, which I was particularly happy about) but that's really not necessary and doesn't add information to the user, so I'd just remove it

antoine-levitt commented:
OK so that's either Infiltrator.store.var or set_store(Main) and have it in the global namespace. I wish there was a middle ground, something like `store.var` with having to prefix by Infiltrator. `store` seems kind of a generic name to export, but would that really be so bad? After all Infiltrator is something you `using` at the REPL, so it's unlikely to clash too much.

pfitzseb commented:
Yeah I'm not super happy about that either, although `mystore = Module(); Infiltrator.set_store(mystore); mystore.var` would of course work.

I'd love to find a short and punny name for `store` along the whole infiltration theme, but nothing really comes to mind.

pdeffebach commented on src/Infiltrator.jl:
```diff
@@ -159,57 +82,102 @@ const TEST_TERMINAL_REF = Ref{Any}(nothing)
+  n = length(get_scratch_pad_names(s))
+  print(io, "Infiltrator scratch pad with $(n) name$(n == 1 ? "" : "s")")
+end
+function Base.getproperty(sp::ScratchPad, s::Symbol)
```
We can use `getfield` or other helper functions to avoid name collisions? the `=== :store` seems a bit fragile.

pdeffebach commented on src/Infiltrator.jl:
```diff
@@ -159,57 +82,102 @@ const TEST_TERMINAL_REF = Ref{Any}(nothing)
+Mainly intended for interactive use, as changes to the scratch pads state will not
+propagate into the returned expression.
+"""
+macro with(ex)
```
Can we rename this something else to avoid a conflict with DataFramesMeta or StaticModules? 

I've filed an issue to for re-naming in DataFramesMeta [here](https://github.com/JuliaData/DataFramesMeta.jl/issues/252) because it's bad practice to squat on such a general name. But at the same time this is also a specialized implementation so it also deserves a `@withstore` or similar.

antoine-levitt commented:
There's safe and cache but both these words are too overloaded in a computing setting unfortunately...

pdeffebach commented on src/Infiltrator.jl:
```diff
@@ -159,57 +82,102 @@ const TEST_TERMINAL_REF = Ref{Any}(nothing)
 
 """
-    get_scratch_pad()
+    set_store(s::ScratchPad = Infiltrator.store, m::Module)
```
Add an exclamation mark? Since it's modifying `s`, but at the same time it will usually be called without the first argument. Not sure what the convention is.

pdeffebach reviewed (commented):
Thanks for this! My only comments are minor renaming things. 

I think we should bike-shed `store` a little as well, since its generally a verb and not used as a noun. Something like suitcase or briefcase? Spies carry briefcases.

pfitzseb commented:
Thoughts on `hoard` or `pile`?

pdeffebach commented:
`hoard` has the same problem that it's generally a verb. I'm fine with `pile` since it emphasizes that this is kind of a messy dumping ground for variables. Downside is it's not with the spy theme 😄

pfitzseb commented:
Ok, let's crowd source this on Slack ;)

Addressed most of the review comments so far (thanks both!), but the `store` bike shedding should probably happen before I tag a 1.0.

antoine-levitt commented:
I'm a bit sad about this because of the cognitive burden of having to remember the "safehouse" pun, but at least it's tab-completion friendly.

pfitzseb commented:
Fair, but most of the more obvious names are too common to be exported, imho. And if someone comes up with a great alternative we can always add another exported name for it :)

antoine-levitt commented:
I liked `exfiltrated`, which is fine to export and easier to remember.

pfitzseb commented:
https://github.com/JuliaDebug/Infiltrator.jl/commit/d7bbc598e50cd8278051d1214ab1ced90bb39d48
